platform-sdk/swirlds-state-impl/src/main/java/com/swirlds/state/merkle/disk/OnDiskIterator.java

```diff
@@ -12,26 +15,34 @@
 import java.util.NoSuchElementException;
 
 public class OnDiskIterator<K, V> implements Iterator<K> {
+
     private final MerkleIterator<MerkleNode> itr;
+    private final Codec<K> keyCodec;
     private K next = null;
 
-    public OnDiskIterator(@NonNull final VirtualMap<OnDiskKey<K>, OnDiskValue<V>> virtualMap) {
+    public OnDiskIterator(@NonNull final VirtualMap virtualMap, @NonNull final Codec<K> keyCodec) {
         itr = requireNonNull(virtualMap).treeIterator();
+        this.keyCodec = requireNonNull(keyCodec);
     }
 
     @Override
-    @SuppressWarnings("unchecked")
     public boolean hasNext() {
         if (next != null) {
             return true;
         }
         while (itr.hasNext()) {
             final var merkleNode = itr.next();
-            if (merkleNode instanceof VirtualLeafNode<?, ?> leaf) {
+            if (merkleNode instanceof VirtualLeafNode leaf) {
                 final var k = leaf.getKey();
-                if (k instanceof OnDiskKey<?> onDiskKey) {
-                    this.next = (K) onDiskKey.getKey();
+                // VirtualMap metadata should not be considered as a possible result of the iterator
+                if (k.equals(VM_STATE_KEY)) {
+                    continue;
+                }
+                try {
+                    this.next = keyCodec.parse(k);
```

Comment:
There is potential to do the parsing lazily. Not sure if there is a gain or not. It depends on if when iterating every key is used or if many as skipped.